### PR TITLE
ENH: add method that creates new spline for a partial derivative of bivariate spline

### PR DESCRIPTION
**Pull request by congma**
This changeset addresses issue #2047 by adding a method `partial_derivative` to the base class `BivariateSpline`.

<s>The major benefit is increased performance when evaluating derivatives with `grid=False`.</s> The major benefit is increased performance from reduction of call overhead. A typical usage case where this benefit may show up is the repeated evaluation of a particular derivative of the same spline with scalar arguments, for example in the integrand function of numerical quadrature using `scipy.integrate.dblquad` or `scipy.integrate.nquad`. Coefficients for the new spline representing the derivative are evaluated once and reused for each subsequent evaluations.

```Python console
In [1]: %paste                                                                  
import numpy
import scipy.interpolate as spi

xs = numpy.linspace(0, numpy.pi, 100)
ys = xs.copy()
zs = numpy.sin(xs[:, None] + ys[None, :])

orig = spi.RectBivariateSpline(xs, ys, zs)
der = orig.partial_derivative(0, 1)
## -- End pasted text --

In [2]: %timeit orig(0.23, 0.23, 0, 1, grid=False)                              
28.9 µs ± 322 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [3]: %timeit der(0.23, 0.23, grid=False)                                     
3.88 µs ± 45.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

The implementation uses a modified version of the FORTRAN code `parder.f`, named `pardtc.f`, to evaluate the new coefficients. The knot and coefficient arrays are then sliced in the Python code and passed to the `_from_tck` private classmethod to construct a new instance.

Things that I'm unsure of:

- maintainability of duplicated FORTRAN code
- better organization of unit tests for the bivariate splines (duplicated code; might want to do housekeeping in a separate changeset)
- whether it will play nice with typical subclasses of `BivariateSpline`

and I'd like to hear your comments. Thanks!

*edit*: more accurate description for the usage case where this could be beneficial.

### CODE REVIEW

**Comment by congma** [comment]
Current status of this branch:

- Tests should now fully cover the changes introduced.
- Re-factored the class inheritance hierarchy a bit for better accommodation of the `partial_derivative` method, without affecting the typical usage of all other classes under `scipy.interpolate.fitpack2`.
- Docstrings have covered all the changes introduced.

I will leave the branch as it currently is for a while and welcome your attention & review. The last rebase and force-push were to resolve a problem with outdated `master` in my repo affecting the sphinx build on CircleCI.

**Comment by rgommers** [comment]
Thanks @congma, this looks interesting. We're a bit light on reviewers for Fortran code at the moment, so this could take a little while to review - apologies in advance.

**Comment by congma** [comment]
@rgommers, there's nothing to apologize for!

File: scipy/interpolate/fitpack/pardtc.f [hunk 1]
```diff
@@ -0,0 +1,155 @@
+      subroutine pardtc(tx,nx,ty,ny,c,kx,ky,nux,nuy,newc,ier)
```
Comment (ev-br) [inline scipy/interpolate/fitpack/pardtc.f:1]: Maybe add `recursive subroutine`, cf gh-16053

File: scipy/interpolate/fitpack/pardtc.f [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+c  subroutine pardtc takes the knots and coefficients of a bivariate
```
Comment (ev-br) [inline scipy/interpolate/fitpack/pardtc.f:2]: Minor: consider adding `implicit none`

File: scipy/interpolate/fitpack2.py [hunk 3]
```diff
@@ -932,6 +943,55 @@ def __call__(self, x, y, dx=0, dy=0, grid=True):
...
+            derivative of this spline.
+
+        Notes
+        -----
+        The returned spline is an instance of the class
+        ``_DerivedBivariateSpline`` that supports spline evaluation operations,
+        but is not directly constructed from original data.
+        """
+        if nux == 0 and nuy == 0:
+            return self
```
Comment (ev-br) [inline scipy/interpolate/fitpack2.py:970]: Just thinking out loud: non-zero values create a copy , while this returns essentially a view. Could this be problematic? Cannot come up with a test case right now, so maybe it's FUD.

File: scipy/interpolate/fitpack2.py [hunk 4]
```diff
@@ -997,6 +1057,7 @@ class BivariateSpline(_BivariateSplineBase):
         a bivariate spline using weighted least-squares fitting
     RectSphereBivariateSpline :
         a bivariate spline over a rectangular mesh on a sphere
+        to create a BivariateSpline using weighted least-squares fitting
```
Comment (ev-br) [inline scipy/interpolate/fitpack2.py:1060]: A leftover from earlier iterations?

File: scipy/interpolate/tests/test_fitpack2.py [hunk 5]
```diff
@@ -1,4 +1,7 @@
 # Created by Pearu Peterson, June 2003
+from __future__ import division, print_function, absolute_import
```
Comment (ev-br) [inline scipy/interpolate/tests/test_fitpack2.py:2]: This line may go

File: scipy/interpolate/tests/test_fitpack2.py [hunk 6]
```diff
@@ -859,6 +862,47 @@ def test_derivatives(self):
...
+            [2,.25,0,-.25,-2],[4,-1,0,1,-4]])
+        dxdy = array([[40,-25,0,25,-40],[-26,16.25,0,-16.25,26],
+            [-8,5,0,-5,8],[22,-13.75,0,13.75,-22],[-8,5,0,-5,8]])/6.
+        lut = RectBivariateSpline(x,y,z)
+        assert_array_almost_equal(lut.partial_derivative(1, 0)(x, y), dx)
+        assert_array_almost_equal(lut.partial_derivative(0, 1)(x, y), dy)
+        assert_array_almost_equal(lut.partial_derivative(1, 1)(x, y), dxdy)
+
+    def test_partial_derivative_method(self):
+        x = array([1,2,3,4,5])
```
Comment (ev-br) [inline scipy/interpolate/tests/test_fitpack2.py:881]: The linter will complain about spaces after commas. Can be ignored (will only complain once AFAIU)

File: scipy/interpolate/tests/test_fitpack2.py [hunk 7]
```diff
@@ -859,6 +862,47 @@ def test_derivatives(self):
...
+        y = array([1,2,3,4,5])
+        z = array([[1,2,1,2,1],[1,2,1,2,1],[1,2,3,2,1],[1,2,2,2,1],[1,2,1,2,1]])
+        dx = array([0,0,2./3,0,0])
+        dy = array([4,-1,0,-.25,-4])
+        dxdy = array([160,65,0,55,32])/24.
+        lut = RectBivariateSpline(x,y,z)
+        assert_array_almost_equal(lut.partial_derivative(1, 0)(x, y,
+                                                               grid=False),
+                                  dx)
```
Comment (ev-br) [inline scipy/interpolate/tests/test_fitpack2.py:890]: Would be nice to add a comment on the origin of these numbers, here and below.

File: scipy/interpolate/fitpack/pardtc.f [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+c     call fpbisp(tx(nux+1),nx-2*nux,ty(nuy+1),ny-2*nuy,newc,newkx,newky,
+c    * x,mx,y,my,z,newc(iwx),newc(iwy),iwrk(1),iwrk(mx+1))
+c
+c from bispev.f:
+c     call fpbisp(tx,       nx,      ty,       ny,      c,   kx,   ky,
+c    * x,mx,y,my,z,wrk(1),   wrk(iw),  iwrk(1),iwrk(mx+1))
+c
+c from fpbisp.f:
+c          fpbisp(tx,       nx,      ty,       ny,      c,   kx,   ky,
+c    * x,mx,y,my,z,wx,       wy,       lx,     ly)
```
Comment (ev-br) [inline scipy/interpolate/fitpack/pardtc.f:152]: These comments may go or do they serve a purpose? Please add a comment if it's the latter.

**Review by ev-br — COMMENTED** [review]
First and foremost, apologies for letting this slip and sit for this long @congma !

The PR looks great overall! I've added a couple of comments, mostly on cosmetics, none are blocking. Would be nice to address them if you can, otherwise I'll do it. 
The `pardtc` routine looks authentic, would not have guessed it's not from Dierckx himself :-). 

On a side note, would be great if you could weigh in on gh-16053. And there are several other things in flight now where help from a fortan person would be superb.

**Comment by ev-br** [comment]
Continued in https://github.com/scipy/scipy/pull/16220

**Pull request closed without merging**